tests/phpunit/tests/post/thumbnails.php

```diff
@@ -450,6 +450,38 @@ public function data_post_thumbnail_size_filter_complex() {
 		);
 	}
 
+	/**
+	 * @ticket 58212
+	 * @covers ::_wp_post_thumbnail_context_filter_add
+	 * @covers ::_wp_post_thumbnail_context_filter_remove
+	 * @covers ::_wp_post_thumbnail_context_filter
+	 */
+	public function test_get_the_post_thumbnail_uses_correct_context() {
+		$last_context = '';
+		add_filter(
+			'wp_get_attachment_image_context',
+			function( $context ) use ( &$last_context ) {
+				$last_context = $context;
+				return $context;
+			},
+			11
+		);
+
+		set_post_thumbnail( self::$post, self::$attachment_id );
+
+		// Assert that before changing the context via filter it's the regular context.
+		wp_get_attachment_image( self::$attachment_id );
+		$this->assertSame( 'wp_get_attachment_image', $last_context, 'Incorrect context for wp_get_attachment_image() before calling get_the_post_thumbnail()' );
+
+		// Assert that get_the_post_thumbnail() triggers a temporary filter to override the context.
+		get_the_post_thumbnail( self::$post );
+		$this->assertSame( 'the_post_thumbnail', $last_context, 'Incorrect context for get_the_post_thumbnail()' );
+
+		// Assert that the regular context is used again following the temporary filter being removed.
+		wp_get_attachment_image( self::$attachment_id );
+		$this->assertSame( 'wp_get_attachment_image', $last_context, 'Incorrect context for wp_get_attachment_image() after calling get_the_post_thumbnail()' );
+	}
```

Comment:
```suggestion
    /**
	 * @ticket 58212
	 * @covers ::_wp_post_thumbnail_context_filter_add
	 * @covers ::_wp_post_thumbnail_context_filter_remove
	 * @covers ::_wp_post_thumbnail_context_filter
	 */
	public function test_get_the_post_thumbnail_uses_correct_context_before_filter() {
		$last_context = '';
		add_filter(
			'wp_get_attachment_image_context',
			function( $context ) use ( &$last_context ) {
				$last_context = $context;
				return $context;
			},
			11
		);

		set_post_thumbnail( self::$post, self::$attachment_id );

		// Assert that before changing the context via filter it's the regular context.
		wp_get_attachment_image( self::$attachment_id );
		$this->assertSame( 'wp_get_attachment_image', $last_context, 'Incorrect context for wp_get_attachment_image() before calling get_the_post_thumbnail()' );
	}

	/**
	 * @ticket 58212
	 * @covers ::_wp_post_thumbnail_context_filter_add
	 * @covers ::_wp_post_thumbnail_context_filter_remove
	 * @covers ::_wp_post_thumbnail_context_filter
	 */
	public function test_get_the_post_thumbnail_triggers_temporary_filter() {
		$last_context = '';
		add_filter(
			'wp_get_attachment_image_context',
			function( $context ) use ( &$last_context ) {
				$last_context = $context;
				return $context;
			},
			11
		);

		set_post_thumbnail( self::$post, self::$attachment_id );

		wp_get_attachment_image( self::$attachment_id );

		// Assert that get_the_post_thumbnail() triggers a temporary filter to override the context.
		get_the_post_thumbnail( self::$post );
		$this->assertSame( 'the_post_thumbnail', $last_context, 'Incorrect context for get_the_post_thumbnail()' );
	}

	/**
	 * @ticket 58212
	 * @covers ::_wp_post_thumbnail_context_filter_add
	 * @covers ::_wp_post_thumbnail_context_filter_remove
	 * @covers ::_wp_post_thumbnail_context_filter
	 */
	public function test_get_the_post_thumbnail_uses_correct_context_after_filter() {
		$last_context = '';
		add_filter(
			'wp_get_attachment_image_context',
			function( $context ) use ( &$last_context ) {
				$last_context = $context;
				return $context;
			},
			11
		);

		set_post_thumbnail( self::$post, self::$attachment_id );
		wp_get_attachment_image( self::$attachment_id );
		get_the_post_thumbnail( self::$post );

		// Assert that the regular context is used again following the temporary filter being removed.
		wp_get_attachment_image( self::$attachment_id );
		$this->assertSame( 'wp_get_attachment_image', $last_context, 'Incorrect context for wp_get_attachment_image() after calling get_the_post_thumbnail()' );
	}
```

We can break this out to three different function.

Reply:
Thanks for the ping @felixarntz!

There can be multiple assertions in a test, but there should generally only be one thing being tested.

For example:
```php
$actual_posts = get_posts( $test_args );
$this->assertIsArray( $actual_posts, 'Posts were not retrieved.' );
$this->assertCount( 2, $actual_posts, 'The wrong number of posts were returned.' );
$this->assertSame( 1, $actual_posts[0]->ID, 'The first post has the wrong ID.' );
$this->assertSame( 2, $actual_posts[1]->ID, 'The second post has the wrong ID.' );
```
This test would be checking the IDs of the first two posts returned by a query. It requires four assertions to be stable. If it were testing multiple queries, it should use a data provider. If it were testing the same query with varying environment conditions, it should probably be different tests.

Back to the PR though. It does look like three different things are being tested. To reduce duplication, we could abstract `$last_context` and the `add_filter()` call to a helper method that returns `$last_context`.

In addition, we can test each new function independently in unit tests, as well as integration tests.

This ensures that:
- The functions work as intended.
- The integration of the functions works as intended.
- If the integration tests fail and the unit tests pass, then the problem lies in the integration.
- If the integration tests fail and the unit tests fail, then the problem lies in the functions.

The tests should cover the following functions:
- `wp_get_attachment_image()` - BC: Behaves as normal when called directly.
- `_wp_post_thumbnail_context_filter()` - Returns the expected context.
- `_wp_post_thumbnail_context_filter_add()` - Adds a context override.
- `_wp_post_thumbnail_context_filter_remove()` - Removes a context override.
- `get_the_post_thumbnail()`
  - Uses the `'the_post_thumbnail'` context.
  - Restores the original context afterwards.

Suggested tests:
```php
/**
 * Tests that `wp_get_attachment_image()` uses the 'wp_get_attachment_image' context by default.
 *
 * @ticket 58212
 *
 * @covers ::wp_get_attachment_image
 */
public function test_wp_get_attachment_image_uses_correct_context_by_default() {
	$this->helper_get_last_context( true );

	set_post_thumbnail( self::$post, self::$attachment_id );
	wp_get_attachment_image( self::$attachment_id );

	$this->assertSame( 'wp_get_attachment_image', $this->helper_get_last_context() );
}

/**
 * Tests that `_wp_post_thumbnail_context_filter()` returns 'the_post_thumbnail'.
 *
 * @ticket 58212
 *
 * @covers::_wp_post_thumbnail_context_filter
 */
public function test_wp_post_thumbnail_context_filter_should_return_the_post_thumbnail() {
	$this->assertSame( 'the_post_thumbnail', _wp_post_thumbnail_context_filter( 'wp_get_attachment_image' ) );
}

/**
 * Tests that `::_wp_post_thumbnail_context_filter_add` adds a filter to override the context
 * used in `wp_get_attachment_image()`.
 *
 * @ticket 58212
 *
 * @covers ::_wp_post_thumbnail_context_filter_add
 */
public function test_wp_post_thumbnail_context_filter_add_should_add_the_filter() {
	$this->helper_get_last_context( true );

	set_post_thumbnail( self::$post, self::$attachment_id );
	_wp_post_thumbnail_context_filter_add();
	wp_get_attachment_image( self::$attachment_id );

	$this->assertSame( 'the_post_thumbnail', $this->helper_get_last_context() );
}

/**
 * Tests that `_wp_post_thumbnail_context_filter_remove()` removes a filter to override the context
 * used in `wp_get_attachment_image()`.
 *
 * @ticket 58212
 *
 * @covers ::_wp_post_thumbnail_context_filter_remove
 */
public function test_wp_post_thumbnail_context_filter_remove_should_remove_the_filter() {
	$this->helper_get_last_context( true );

	set_post_thumbnail( self::$post, self::$attachment_id );
	_wp_post_thumbnail_context_filter_add();
	wp_get_attachment_image( self::$attachment_id );

	// Verify that the filter has been added before testing that it has been removed.
	$this->assertSame(
		'the_post_thumbnail',
		$this->helper_get_last_context(),
		'The filter was not added.'
	);

	_wp_post_thumbnail_context_filter_remove();

	// This should no longer be modified by the filter.
	wp_get_attachment_image( self::$attachment_id );

	$this->assertSame(
		'wp_get_attachment_image',
		$this->helper_get_last_context(),
		'The filter was not removed.'
	);
}

/**
 * Tests that `get_the_post_thumbnail()` uses the 'the_post_thumbnail' context.
 *
 * @ticket 58212
 *
 * @covers ::get_the_post_thumbnail
 */
public function test_get_the_post_thumbnail_should_use_the_post_thumbnail_context() {
	$this->helper_get_last_context( true );

	set_post_thumbnail( self::$post, self::$attachment_id );
	get_the_post_thumbnail( self::$post );

	$this->assertSame( 'the_post_thumbnail', $this->helper_get_last_context() );
}

/**
 * Tests that `get_the_post_thumbnail()` restores the context afterwards.
 *
 * @ticket 58212
 *
 * @covers ::get_the_post_thumbnail
 */
public function test_get_the_post_thumbnail_should_remove_the_post_thumbnail_context_afterwards() {
	$this->helper_get_last_context( true );

	set_post_thumbnail( self::$post, self::$attachment_id );
	get_the_post_thumbnail( self::$post );

	// Verify that the context was overridden before testing that it has been restored.
	$this->assertSame(
		'the_post_thumbnail',
		$this->helper_get_last_context(),
		'The context was not overridden.'
	);

	wp_get_attachment_image( self::$attachment_id );

	$this->assertSame(
		'wp_get_attachment_image',
		$this->helper_get_last_context(),
		'The context was not restored.'
	);
}

/**
 * Helper method to get the last context returned by the 'wp_get_attachment_image' filter.
 *
 * @param bool $reset Whether to reset the context to an empty string.
 *
 * @return string The last context.
 */
private function helper_get_last_context( $reset = false ) {
	static $last_context = '';

	if ( $reset ) {
		$last_context = '';
		add_filter(
			'wp_get_attachment_image_context',
			function( $context ) use ( &$last_context ) {
				$last_context = $context;
				return $context;
			},
			11
		);
	}

	return $last_context;
}
```

Reply:
Thanks @costdev for the detailed explanation and code example. I've updated the test code to use separate tests, though note that the first test is actually already covered in the `media.php` test class, so that's another good point that we don't even need that assertion here.

The main thing I solved slightly differently is the helper method, I think we can address that in a cleaner way without coupling to a static property and involving two methods by instead passing a variable by reference to keep track of the value within a single test.

Can you please take another look?

Reply:
Yep, that looks good to me! Nice one!